### ISSUE TITLE
[HOLD for payment 2023-10-23] [HOLD for payment 2023-10-23] [HOLD for payment 2023-10-20] [HOLD for payment 2023-10-20] [$2000] Unable to change the currency after double-clicking on the current currency

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#22497]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Log in to your account.
2. Click on the FAB menu and select either the "Request Money" or "Split Bill" option.
3. Double-click on the current currency.
4. In the Currency list, select a different currency.
5. Observe that the selected currency does not change.
## Expected Result: 
The currency should change to the selected one.
## Actual Result: 
The selected currency remains unchanged after double-clicking on the current currency.

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [x] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.38-3
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/6a06fd67-b83f-4784-a0d6-586fcedb8f3b

https://github.com/Expensify/App/assets/43996225/7296c3ad-ad0c-42a3-953e-04c576205cb4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @tranvantoan-qn
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1688821290443479

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01ac5f2b1e902a7171</li>
        <li>Upwork Job ID: 1678505607142854656</li>
        <li>Last Price Increase: 2023-08-06</li>
    </ul>
</details>

**Comment by maddylewis** [comment]
bumping @eVoloshchak to review existing proposals. lmk if you're overloaded and i can reach out to another C+ - thx!

**Comment by eVoloshchak** [comment]
@ginsuma, thanks for the [proposal](https://github.com/Expensify/App/issues/22497#issuecomment-1627487929)!

Debouncing navigate would _technically_ resolve the issue, but I think that's more of a workaround. Let's resolve the root cause instead

@bernhardoj, can we implement the same logic in `PressableWithoutFeedback` instead? (ideally, shouldn't be this implemented in `GenericPressable` component, so all Pressables support this behavior?)

**Comment by priyeshshah11** [comment]
> > ideally, shouldn't be this implemented in GenericPressable component,
> 
> That's what I thought too, that's why I wrote
> 
> > idk the reason the press logic is only applied to PressableWithFeedback
> 
> in my proposal 😄
> 
> @priyeshshah11 @s77rt maybe can help answer it.

even I don't know the actual reason why we just implemented it in `PressableWithFeedback` & not in a more generic component but that's what was suggested by @roryabraham/@robertKozik at that time, better to check with them.

**Comment by tranvantoan-qn** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Unable to change the currency after double-clicking on the current currency

### What is the root cause of that problem?
The navigation to the Currency selection page is handled in the following code.
https://github.com/Expensify/App/blob/67bac9d50b9fbc1b8a1b1a14e8cc567ca3d0580b/src/pages/iou/steps/MoneyRequestAmountPage.js#L361-L365

In normal cases when the currency is clicked, the `activeRoute`` will be as follows:

- Send Money: /send/new
- Split Money: /split/new
- Request Money: /request/new

But when the current currency is double-clicked, it somehow adds `/currency/`` to the end and an error occurs.

- Send Money: /send/new/currency/
- Split Money: /split/new/currency/
- Request Money: /request/new/currency/

### What changes do you think we should make in order to solve the problem?
Add a condition to check if the `activeRoute` contains `/currency`. Only navigate if `/currency` is not present


```javascript

    const navigateToCurrencySelectionPage = () => {
        // Remove query from the route
        const activeRoute = Navigation.getActiveRoute().replace(/\?.*/, '')
        if (!activeRoute.includes('/currency')) {
            // Encode the route.
            const backToRoute = encodeURIComponent(activeRoute);
            Navigation.navigate(ROUTES.getMoneyRequestCurrencyRoute(iouType.current, reportID.current, selectedCurrencyCode, backToRoute));
        }
    };

```

### What alternative solutions did you explore? (Optional)
N/A

**Comment by tranvantoan-qn** [comment]
@eVoloshchak 
I don't think adding Debouncing will help
I've added my proposal [here](https://github.com/Expensify/App/issues/22497#issuecomment-1637658183), please have a look.

**Comment by maddylewis** [comment]
@eVoloshchak - bump on this one - ty!

**Comment by eVoloshchak** [comment]
@tranvantoan-qn, [this](https://github.com/Expensify/App/issues/22497#issuecomment-1637658183) is a workaround, let's prevent this function from being called twice in the first place

@bernhardoj, I think we should implement this in the generic pressable, so it's universally available
Would you mind posting an updated proposal?

**Comment by eVoloshchak** [comment]
@bernhardoj's updated [proposal](https://github.com/Expensify/App/issues/22497#issuecomment-1627690548) looks good to me
Let's make sure this is fixed for all pressables by implementing the alternative solution
> We can move the press logic above from PressbaleWithFeedback to BaseGenericPressable, so both WithFeedback and WithoutFeedback will have the same logic

🎀👀🎀 C+ reviewed!

**Comment by francoisl** [comment]
@bernhardoj's [proposal](https://github.com/Expensify/App/issues/22497#issuecomment-1627690548) looks like the best to me as well, let's go with that.

**Comment by maddylewis** [comment]
thanks for the update @bernhardoj - once @eVoloshchak and @francoisl review, we'll move forward with the fix. thanks!

**Comment by francoisl** [comment]
@bernhardoj it makes sense to me to move that logic up to `BaseGenericPressable`, though in the end I don't see how that would work any better than your original proposal. What exactly keeps causing an issue with your original proposal?

**Comment by eVoloshchak** [comment]
@bernhardoj, the [updated proposal ](https://github.com/Expensify/App/issues/22497#issuecomment-1627690548)does resolve the issue, but it's a workaround, navigate is still called twice, just with the correct route.
Is there a way we could disable `Pressable` while navigation action is in progress?

**Comment by maddylewis** [comment]
@eVoloshchak - can you provide an update on where we are with this fix? ty!

**Comment by maddylewis** [comment]
I am OOO until ~Aug 20 - here is the latest update for this one - https://github.com/Expensify/App/issues/22497#issuecomment-1663990410

it sounds like this isn't "critical" enough to warrant moving forward with the existing proposals. and we should wait for some new proposals. 

i will double the bounty on this one, though!

**Comment by pradeepmdk** [comment]
### Proposal
### Please re-state the problem that we are trying to solve in this issue.
Unable to change the currency after double-clicking on the current currency

### What is the root cause of that problem?
https://github.com/Expensify/App/blob/014b326a1cb193448746a7fec90cf0411563a35a/src/pages/iou/steps/NewRequestAmountPage.js?#L143
`navigateToCurrencySelectionPage` called twice on the double tab so each time navigation state is changed 
https://github.com/Expensify/App/blob/014b326a1cb193448746a7fec90cf0411563a35a/src/pages/iou/steps/NewRequestAmountPage.js?#L145


### What changes do you think we should make in order to solve the problem?

1) prevent the double tab (this is already proposed)
2) we can modify the active route variable instead of getting active route
  
```javascript
   const navigateToCurrencySelectionPage = () => {
        // Remove query from the route and encode it.
        const activeRoute = encodeURIComponent(ROUTES.getMoneyRequestRoute(iouType, reportID)) //encodeURIComponent(Navigation.getActiveRoute().replace(/\?.*/, ''));
        Navigation.navigate(ROUTES.getMoneyRequestCurrencyRoute(iouType, reportID, currency, activeRoute));
    };

``` 
this will help us maintain same path in the state

**Comment by stephanieelliott** [comment]
Some more proposals [here](https://github.com/Expensify/App/issues/22497#issuecomment-1667019381) and [here](https://github.com/Expensify/App/issues/22497#issuecomment-1667555128), @eVoloshchak can you please review?

**Comment by pradeepmdk** [comment]
@eVoloshchak  any update

**Comment by eVoloshchak** [comment]
Reviewing shortly

**Comment by eVoloshchak** [comment]
> prevent the double tab (using throttle)

@pradeepmdk, this was proposed [before](https://github.com/Expensify/App/issues/22497#issuecomment-1627487929), while would technically work, this is working around the problem 
> we can modify the active route variable instead of getting getActiveRoute

A similar approach was proposed before (https://github.com/Expensify/App/issues/22497#issuecomment-1627690548), while your implementation is a bit different, I think the general idea is the same

**Comment by eVoloshchak** [comment]
> We should add an isEditingCurrency variable that we set to true when onCurrencyButtonPress inside navigateToCurrencySelectionPage and clear it when we navigate back to the request money page.

@WikusKriek, while this would work, I wonder if there's a more universal approach. This problem can be observed everywhere throughout the app, every button that triggers navigation will have the same problem. For instance, pressing back button twice will `navigate back` twice (notice the animation)

https://github.com/Expensify/App/assets/9059945/d5ff4a71-cfe3-43dd-aa08-b85acb105b38

Here are our options

1. Extend this issue to fix this bug throughout the app (every button that triggers navigation will have the same problem when double-clicking). In principle, we need to disable all navigation buttons while the navigation transition is in progress. Not sure how complex that would be and what the value is, since mostly the bug is animation being played twice. On the other hand, this is a legitimate bug and it needs to be fixed at some point
2. Implement a fix for this particular case (i.e. fix this particular issue with not being able to change the currency after double-clicking on the current currency). (Get the currently active route once instead of getting it on each click. https://github.com/Expensify/App/issues/22497#issuecomment-1627690548)

@arosiclair, what do you think?

**Comment by pradeepmdk** [comment]
@eVoloshchak  thanks for reviewing my proposal.
why I am suggesting using `ROUTES.getMoneyRequestRoute(iouType, reportID)` means we are already in the `IOUCurrencySelection.js` using this only on goback session. i am not sure why still we need to pass activeroute from `MoneyRequestAmountPage.js` 
https://github.com/Expensify/App/blob/014b326a1cb193448746a7fec90cf0411563a35a/src/pages/iou/IOUCurrencySelection.js?#L140 

and here our main issue on double tab. we want  need only single tab. we deviate two events from pressable component single or double here we can do that
https://github.com/Expensify/App/blob/60f48175c9e3342ef192d878218820e5ed5a76cd/src/components/Pressable/GenericPressable/index.native.js


```javascript
const NativeGenericPressable = forwardRef((props, ref) => {
    const lastTap = useRef(null);

    return (
        <GenericPressable
            focusable
            accessible
            accessibilityHint={props.accessibilityHint || props.accessibilityLabel}
            // eslint-disable-next-line react/jsx-props-no-spreading
            {...props}
            onPress={() => {
                const now = Date.now();
                const DOUBLE_PRESS_DELAY = 300; 
                if (lastTap.current && now - lastTap.current < DOUBLE_PRESS_DELAY) {
                    // Double tap logic
                    console.log('Double tap detected');
                    props.onDoublePress && props.onDoublePress();
                  } else {
                    // Single tap logic
                    console.log('Single tap detected');
                    lastTap.current = now;
                    props.onPress && props.onPress();
                  }
            }}
            ref={ref}
        />
    )
});
``` 
so that we can avoid this in all the places on the double tab.. and also it should be a feature request. if we need anything on the double tab we can use this.

**Comment by pradeepmdk** [comment]
@WikusKriek
In my proposal in the common component I suggested split the events that is single click or double click because by default native will use all the click event pass via onpress.

onPress={_.debounce(props.onPress, 300)}
 For this not a correct solution because this will delay 300ms for single click also 

If we prevent double click we need to use this via debounce
onPress={_.debounce(props.onPress, 300,{
    leading: true,
    trailing: false,
  })}

When set leading true it will trigger immediate and trialing is false so it will prevent the subsequent call in the 300ms

**Comment by pradeepmdk** [comment]
@WikusKriek we have options to pass this.
{
leading: true,
trailing: false,
}
<img width="1129" alt="Screenshot 2023-08-14 at 7 24 42 PM" src="https://github.com/Expensify/App/assets/16595846/5e4761e5-f713-46d1-afa6-1f3b2fc3c616">
<img width="850" alt="Screenshot 2023-08-14 at 7 25 00 PM" src="https://github.com/Expensify/App/assets/16595846/808e6ac2-c4b5-4148-9d12-07a9b62b2626">


```javascript
onPress={_.debounce(props.onPress, 300, true)}
``` 
this will also not work

 When you pass true as an option to Lodash's debounce function, it is treated as { leading: true, trailing: true }. This means that the debounced function will be triggered immediately on the first call (leading edge), and it will also be triggered once after the debounce delay has passed (trailing edge).

**Comment by arosiclair** [comment]
> @arosiclair, what do you think?

I think you meant to mention @francoisl 😄

**Comment by eVoloshchak** [comment]
> I think you meant to mention @francoisl 😄

Ooops, sorry😅

@francoisl, what are your thoughts on [this](https://github.com/Expensify/App/issues/22497#issuecomment-1675173149)?

**Comment by eVoloshchak** [comment]
@WikusKriek, @pradeepmdk, I think `debounce` should be our last option, it isn't resolving the root cause, so there could be issues down the line. For instance, what if the navigation transition takes longer than the debounce period?

**Comment by francoisl** [comment]
@eVoloshchak on [this](https://github.com/Expensify/App/issues/22497#issuecomment-1675173149), yes I agree that ideally we can find a solution that fixes the issue throughout the app, instead of just this particular case.

---

@WikusKriek your options 2 and 3 just above sound like a good lead.

> 2. We can do the same for PressableWithoutFeedback as we do with PressableWithFeedback in terms of having the isExecuting logic in onPress.
> 3. Add the PressableWithFeedback isExecuting logic to [NativeGenericPressable](https://github.com/Expensify/App/blob/60f48175c9e3342ef192d878218820e5ed5a76cd/src/components/Pressable/GenericPressable/index.native.js)

Out of curiosity, do you think there would be any advantage or issue in moving that logic to `GenericPressable` instead?

**Comment by pradeepmdk** [comment]
@francoisl  in the GenericPressable also we can do this. but when I am testing the web the OnPress is triggered once only.

@WikusKriek   option 3 is not working for me still have the issue. 

@eVoloshchak  when we use this debounce

```javascript
_.debounce(props.onPress, 300, {
                leading: true,
                trailing: false,
                })
``` 
it will trigger immediately so navigation is triggered without any time delay. after the subsequent calls will block until 300ms. within the time frame navigation is done.

**Comment by pradeepmdk** [comment]
@WikusKriek  for me as well working.

@eVoloshchak @francoisl can you check my solution as well https://github.com/Expensify/App/issues/22497#issuecomment-1675490472

 we can add that in the `GenericPressable` as well.

**Comment by maddylewis** [comment]
@bernhardoj - are you suggesting we place `#22497` on HOLD to see if the fix for https://github.com/Expensify/App/issues/24261 is live?

**Comment by eVoloshchak** [comment]
@bernhardoj, looks like [PR](https://github.com/Expensify/App/pull/24614) for `useSingleExecution` hook is close to being merged, so let's wait for that 
This [solution](https://github.com/Expensify/App/issues/22497#issuecomment-1681028014) definately looks interesting

**Comment by maddylewis** [comment]
still holding on this PR - https://github.com/Expensify/App/pull/24614

**Comment by maddylewis** [comment]
still awaiting merge - https://github.com/Expensify/App/pull/24614

**Comment by saalimzafar** [comment]
If it's working on first click then why for second click it's a bug?? 
And isn't it against the UIUX laws that we are making something clickable which shouldn't be. What tells that it's clickable and at this point, this way?? It's weird.
The issue is not really an issue but we assumed it and for that reason we are looking for solutions.
Things should be kept simple because simplicity beats complexity.

**Comment by maddylewis** [comment]
still awaiting merge - https://github.com/Expensify/App/pull/24614#issuecomment-1698402694

**Comment by maddylewis** [comment]
@eVoloshchak - now that `#24614` is merged, should we move forward with this solution? https://github.com/Expensify/App/issues/22497#issuecomment-1681028014

**Comment by eVoloshchak** [comment]
@bernhardoj, could you please verify if you're still experiencing this [issue](https://github.com/Expensify/App/issues/22497#issuecomment-1645027294) when using `useSingleExecution`?

**Comment by maddylewis** [comment]
@eVoloshchak - what do you suggest as the best path forward at this point?

**Comment by eVoloshchak** [comment]
@bernhardoj, thank you for keeping the investigation going on this!
> @eVoloshchak But before we hold more, can you help me verify the reproduction steps and the solution above?

Yes, I can reproduce this 100% of time using your steps 
Looks like there are [plans](https://github.com/necolas/react-native-web/issues/2349#issuecomment-1195026834) to resolve this in react-native itself in some distant future 
> Having said that, it sounds like this might be an issue that we could/should fix somewhere in this chain of abstractions rather than in your product code.

But it's been over a year and the issue was closed in favor of using `onLongPress={() => {}}`, so it's the next best thing we can do

**Comment by eVoloshchak** [comment]
> @eVoloshchak - what do you suggest as the best path forward at this point?

@maddylewis, let's hold this for https://github.com/Expensify/App/issues/24654

**Comment by maddylewis** [comment]
ill move to weekly while were holding

**Comment by francoisl** [comment]
I think I'm missing how https://github.com/Expensify/App/issues/24654 could be related, but a fix for it was merged - can we take a look into this again?

**Comment by eVoloshchak** [comment]
> So, I think we should apply the empty long press fix to currency Pressable only. What do you think?

@bernhardoj, ideally we'd want this resolved throughout the app, a more universal approach would be much better I think

> they decided to apply that fix to chrome mWeb only

According to https://github.com/necolas/react-native-web/issues/2349 and https://github.com/necolas/react-native-web/issues/2352 this is happening for Android Chrome only. Is it also a native issue or is it a completely different bug we're facing?

**Comment by eVoloshchak** [comment]
@bernhardoj, thanks for the explanation!

I lean towards the second option (disabling the Pressable with `disabled` prop). When using an empty `onLongPress` we're essentially _working around_ the problem, rather than fixing the root cause (pressable still being enabled). I think it's more reliable and future-proof if we just disable the pressable altogether with `disabled` prop

We do use the same approach for **some** components that are utilizing Pressable, so applying this for this specific case rather than globally is an option, but I'll have to test this a bit more before the final decision. The main reason is this has the disadvantage of not being able to show the not-allowed cursor as you've described in https://github.com/Expensify/App/issues/22413#issuecomment-1625701475 (which we can also work around, but it's npt ideal)

Could you post an updated proposal summarizing all of the changes?

**Comment by eVoloshchak** [comment]
@bernhardoj's [updated proposal](https://github.com/Expensify/App/issues/22497#issuecomment-1627690548) looks good to me!

> I also suggested splitting the PR into 2 PRs. The first PR is to move the useSingleExecution logic to BaseGenericPressable and the second PR is to fix the long press issue.

I agree, both are quite big changes that affect generic pressable (which is used throughout the app), it'll be easier to test and pinpoint problems if we divide the changes

🎀👀🎀 C+ reviewed!

**Comment by maddylewis** [comment]
lol what is happening with this issue? am i good to start processing payments to: 
- @eVoloshchak 
- @bernhardoj 
- @tranvantoan-qn

**Comment by maddylewis** [comment]
how are we looking @bernhardoj? are we still waiting on the 2nd PR? https://github.com/Expensify/App/issues/22497#issuecomment-1776419330

**Comment by eVoloshchak** [comment]
In our discussion, we didn't reach a definite answer to the question "Is this worth fixing?", but we did figure out that this behavior is not caused by our code and can be reproduced even in an empty project, so it's a default behavior
> if you tap and hold and there is no onLongPress handler, so the Pressable will "wait" indefinitely until the finger is lifted, triggering the onPress handler

Given that
1. The original issue (Unable to change the currency after double-clicking on the current currency) is fully resolved by the first PR
2. The second issue (tap and hold) is a default behavior we're trying to work around
3. The second issue is essentially an edge case and I don't think users would intentionally perform an action of tapping and then immediately holding the currency button
4. **The risk of potential regressions and further complication of logic of one of our core components (Pressable) outweighs the value of resolving the said issue**

I think we should just close the [second PR](https://github.com/Expensify/App/pull/29317) and not proceed with the fix further
@chiragsalian, what are your thoughts?

**Comment by chiragsalian** [comment]
Yeah i think discussing the 2nd problem in a separate issue makes sense to me. I like having an issue for a very specific usecase like this because if we don't want to solve it now and we close it that's fine. But if we ever want to tackle it in the future there is an existing issue we can open/refer to continue the discussion as needed.

I think considering the first PR solving the main issue here is good enough for me.

**Comment by eVoloshchak** [comment]
I'll fill out the BZ checklist today

**Comment by maddylewis** [comment]
got it - waiting for @eVoloshchak to fill out the BZ checklist. 

this one has a lot of moving parts so i need a little guidance on how payments are looking for this one. let me know how this is looking: 
- Reporting bonus: $250 to @tranvantoan-qn 
- Contributor: $2k to @bernhardoj 
- C+: $2k to @eVoloshchak

**Comment by eVoloshchak** [comment]
- The PR that introduced the bug has been identified. Link to the PR: there isn't a PR that caused this, this was present in the initial implementations
- The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: N/A
-  A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: I don't think an additional discussion is necessary, this is a fairly simple bug (from the testing perspective)

**Regression Test Proposal**

1. Open `Request Money` page
2. Switch to `Manual` tab
3. Double-click the currency symbol button
4. Select any currency
5. Verify you are navigated back and the currency is updated

**Do we agree 👍 or 👎**

**Comment by maddylewis** [comment]
## Payments
- [ ] Reporting bonus: $250 to @tranvantoan-qn (Upwork)
- [ ] Contributor: $2k to @bernhardoj (Upwork)
- [ ] C+: $2k to @eVoloshchak (NewDot)

**Comment by maddylewis** [comment]
everyone is paid

**Comment by tranvantoan-qn** [comment]
@maddylewis There is a previous contract for this one, should it be closed now? 
https://www.upwork.com/nx/wm/workroom/34271376/overview

**Comment by JmillsExpensify** [comment]
$2,000 payment approved for @eVoloshchak based on summary above.
